File: backend/app/services/analysis_service.py
"""
Service layer for analysis-related business logic.
Handles analysis job creation, status tracking, and results retrieval.
"""

import uuid
from datetime import datetime
from typing import List, Optional, Tuple, Dict, Any
from sqlalchemy.orm import Session

from app.models.analysis import AnalysisResult
from app.models.fact_check import FactCheck
from app.models.transcript import Transcript
from app.schemas.analysis import AnalysisJobResponse, JobStatusResponse, AnalysisResultsResponse
from app.utils.logger import get_logger

logger = get_logger(__name__)


class AnalysisNotFoundError(Exception):
    """Raised when an analysis job is not found in the database."""
    pass


class AnalysisService:
    """
    Service class for handling analysis operations.
    
    Provides methods for creating analysis jobs, tracking status,
    and retrieving results with comprehensive logging.
    """

    def __init__(self, db: Session) -> None:
        """
        Initialize the analysis service.
        
        Args:
            db: Database session for operations
        """
        self.db: Session = db

    def create_analysis_job(self, transcript_id: uuid.UUID) -> AnalysisJobResponse:
        """
        Create a new analysis job for a transcript.
        
        Args:
            transcript_id: UUID of the transcript to analyze
            
        Returns:
            AnalysisJobResponse with job details
            
        Raises:
            AnalysisNotFoundError: If transcript is not found
        """
        # Verify transcript exists
        transcript = self.db.query(Transcript).filter(Transcript.id == transcript_id).first()
        if not transcript:
            logger.error("Transcript not found for analysis", transcript_id=transcript_id)
            raise AnalysisNotFoundError(f"Transcript {transcript_id} not found")

        # Generate job ID
        job_id = uuid.uuid4()

        # Create analysis record
        analysis = AnalysisResult(
            transcript_id=transcript_id,
            job_id=job_id,
            status="pending"
        )

        try:
            self.db.add(analysis)
            self.db.commit()
            self.db.refresh(analysis)

            logger.info("Analysis job created", 
                       job_id=job_id, 
                       transcript_id=transcript_id,
                       analysis_id=analysis.id)

            return AnalysisJobResponse(
                job_id=job_id,
                transcript_id=transcript_id,
                status="pending",
                message="Analysis job created and queued for processing"
            )

        except Exception as e:
            self.db.rollback()
            logger.error("Failed to create analysis job", 
                        transcript_id=transcript_id, 
                        error=str(e))
            raise

    def get_job_status(self, job_id: uuid.UUID) -> JobStatusResponse:
        """
        Get the status of an analysis job.
        
        Args:
            job_id: UUID of the analysis job
            
        Returns:
            JobStatusResponse with current job status
            
        Raises:
            AnalysisNotFoundError: If job is not found
        """
        analysis = self.db.query(AnalysisResult).filter(AnalysisResult.job_id == job_id).first()
        if not analysis:
            logger.error("Analysis job not found", job_id=job_id)
            raise AnalysisNotFoundError(f"Analysis job {job_id} not found")

        logger.info("Retrieved job status", 
                   job_id=job_id, 
                   status=analysis.status,
                   analysis_id=analysis.id)

        return JobStatusResponse.from_orm(analysis)

    def update_job_status(self, job_id: uuid.UUID, status: str, error_message: str = None) -> None:
        """
        Update the status of an analysis job.
        
        Args:
            job_id: UUID of the analysis job
            status: New status (pending, processing, completed, failed)
            error_message: Optional error message if status is failed
        """
        analysis = self.db.query(AnalysisResult).filter(AnalysisResult.job_id == job_id).first()
        if not analysis:
            logger.error("Analysis job not found for status update", job_id=job_id)
            return

        analysis.status = status
        if error_message:
            analysis.error_message = error_message
        if status in ["completed", "failed"]:
            analysis.completed_at = datetime.utcnow()

        try:
            self.db.commit()
            logger.info("Updated job status", 
                       job_id=job_id, 
                       status=status,
                       has_error=bool(error_message))

        except Exception as e:
            self.db.rollback()
            logger.error("Failed to update job status", 
                        job_id=job_id, 
                        status=status, 
                        error=str(e))

    def save_analysis_results(self, job_id: uuid.UUID, summary: str, takeaways: List[str]) -> AnalysisResult:
        """
        Save the analysis results (summary and takeaways) to the database.
        
        Args:
            job_id: UUID of the analysis job
            summary: Generated summary text
            takeaways: List of key takeaway strings
            
        Returns:
            Updated AnalysisResult model instance
            
        Raises:
            AnalysisNotFoundError: If job is not found
        """
        analysis = self.db.query(AnalysisResult).filter(AnalysisResult.job_id == job_id).first()
        if not analysis:
            logger.error("Analysis job not found for results save", job_id=job_id)
            raise AnalysisNotFoundError(f"Analysis job {job_id} not found")

        analysis.summary = summary
        analysis.takeaways = takeaways

        try:
            self.db.commit()
            self.db.refresh(analysis)
            
            logger.info("Saved analysis results", 
                       job_id=job_id,
                       analysis_id=analysis.id,
                       summary_length=len(summary),
                       takeaways_count=len(takeaways))

            return analysis

        except Exception as e:
            self.db.rollback()
            logger.error("Failed to save analysis results", 
                        job_id=job_id, 
                        error=str(e))
            raise

    def save_fact_checks(self, analysis_id: uuid.UUID, fact_checks: List[Dict[str, Any]]) -> List[FactCheck]:
        """
        Save fact-check results to the database.
        
        Args:
            analysis_id: UUID of the analysis record
            fact_checks: List of fact-check result dictionaries
            
        Returns:
            List of created FactCheck model instances
        """
        created_fact_checks = []

        for fact_check_data in fact_checks:
            fact_check = FactCheck(
                analysis_id=analysis_id,
                claim=fact_check_data["claim"],
                verdict=fact_check_data["verdict"],
                confidence=fact_check_data["confidence"],
                evidence=fact_check_data.get("evidence"),
                sources=fact_check_data.get("sources", [])
            )
            
            self.db.add(fact_check)
            created_fact_checks.append(fact_check)

        try:
            self.db.commit()
            
            for fact_check in created_fact_checks:
                self.db.refresh(fact_check)

            logger.info("Saved fact-check results", 
                       analysis_id=analysis_id,
                       fact_checks_count=len(created_fact_checks))

            return created_fact_checks

        except Exception as e:
            self.db.rollback()
            logger.error("Failed to save fact-check results", 
                        analysis_id=analysis_id, 
                        error=str(e))
            raise

    def get_analysis_results(self, analysis_id: uuid.UUID) -> AnalysisResultsResponse:
        """
        Get complete analysis results including fact-checks.
        
        Args:
            analysis_id: UUID of the analysis record
            
        Returns:
            AnalysisResultsResponse with complete results
            
        Raises:
            AnalysisNotFoundError: If analysis is not found
        """
        # Join with transcript to get filename and metadata
        result = (self.db.query(AnalysisResult, Transcript)
                 .join(Transcript, AnalysisResult.transcript_id == Transcript.id)
                 .filter(AnalysisResult.id == analysis_id)
                 .first())
        
        if not result:
            logger.error("Analysis not found", analysis_id=analysis_id)
            raise AnalysisNotFoundError(f"Analysis {analysis_id} not found")

        analysis, transcript = result

        logger.info("Retrieved analysis results", 
                   analysis_id=analysis_id,
                   status=analysis.status,
                   fact_checks_count=len(analysis.fact_checks))

        # Create response with transcript information
        response_data = {
            'id': analysis.id,
            'job_id': analysis.job_id,
            'transcript_id': analysis.transcript_id,
            'status': analysis.status,
            'summary': analysis.summary,
            'takeaways': analysis.takeaways,
            'created_at': analysis.created_at,
            'completed_at': analysis.completed_at,
            'transcript_filename': transcript.filename,
            'transcript_title': None,  # Extract from metadata if needed
            'fact_checks': [
                {
                    'id': fc.id,
                    'claim': fc.claim,
                    'verdict': fc.verdict,
                    'confidence': fc.confidence,
                    'evidence': fc.evidence,
                    'sources': fc.sources,
                    'checked_at': fc.checked_at
                }
                for fc in analysis.fact_checks
            ]
        }

        # Extract title from transcript metadata if available
        if transcript.transcript_metadata and isinstance(transcript.transcript_metadata, dict):
            response_data['transcript_title'] = transcript.transcript_metadata.get('title')

        return AnalysisResultsResponse(**response_data)

    def get_analysis_by_job_id(self, job_id: uuid.UUID) -> Optional[AnalysisResult]:
        """
        Get analysis record by job ID.
        
        Args:
            job_id: UUID of the analysis job
            
        Returns:
            AnalysisResult model instance or None if not found
        """
        return self.db.query(AnalysisResult).filter(AnalysisResult.job_id == job_id).first()

    def list_analysis_results(self, skip: int = 0, limit: int = 100) -> Tuple[List[Dict[str, Any]], int]:
        """
        List analysis results with transcript information and pagination.
        
        Args:
            skip: Number of records to skip
            limit: Maximum number of records to return
            
        Returns:
            Tuple of (enriched_analysis_list, total_count)
        """
        # Join analysis results with transcript data
        query = (self.db.query(AnalysisResult, Transcript)
                 .join(Transcript, AnalysisResult.transcript_id == Transcript.id)
                 .order_by(AnalysisResult.created_at.desc()))
        
        total = query.count()
        results = query.offset(skip).limit(limit).all()
        
        # Enrich analysis results with transcript information
        enriched_analyses = []
        for analysis, transcript in results:
            # Extract title from transcript metadata or use filename
            title = None
            if transcript.transcript_metadata and isinstance(transcript.transcript_metadata, dict):
                title = transcript.transcript_metadata.get('title')
            
            # Load fact checks for this analysis
            fact_checks = self.db.query(FactCheck).filter(FactCheck.analysis_id == analysis.id).all()
            fact_checks_data = [
                {
                    'id': fc.id,
                    'claim': fc.claim,
                    'verdict': fc.verdict,
                    'confidence': fc.confidence,
                    'evidence': fc.evidence,
                    'sources': fc.sources,
                    'checked_at': fc.checked_at
                }
                for fc in fact_checks
            ]
            
            # Create enriched analysis object
            analysis_dict = {
                'id': analysis.id,
                'job_id': analysis.job_id,
                'transcript_id': analysis.transcript_id,
                'status': analysis.status,
                'summary': analysis.summary,
                'takeaways': analysis.takeaways,
                'created_at': analysis.created_at,
                'completed_at': analysis.completed_at,
                'transcript_filename': transcript.filename,
                'transcript_title': title or transcript.filename,  # Fallback to filename
                'fact_checks': fact_checks_data
            }
            enriched_analyses.append(analysis_dict)
        
        logger.info("Listed analysis results", 
                   count=len(enriched_analyses), 
                   total=total, 
                   skip=skip, 
                   limit=limit)
        
        return enriched_analyses, total

File: internal/services/analysis_service.go
package services

import (
	"encoding/json"
	"fmt"
	"backend-golang/internal/config"
	"backend-golang/internal/models"
	"backend-golang/pkg/logger"
	"time"

	"github.com/google/uuid"
	"gorm.io/gorm"
)

// AnalysisServiceInterface defines the interface for analysis service operations
type AnalysisServiceInterface interface {
	CreateAnalysisJob(req *AnalysisJobRequest, correlationID string) (*AnalysisJobResponse, error)
	GetJobStatus(jobID uuid.UUID, correlationID string) (*JobStatusResponse, error)
	ListAnalysisResults(page, perPage int) ([]*AnalysisResultsResponse, int64, error)
	GetAnalysisResults(analysisID uuid.UUID, correlationID string) (*AnalysisResultsResponse, error)
	UpdateJobStatus(jobID uuid.UUID, status string, errorMessage string) error
}

// KafkaServiceInterface defines the interface for Kafka operations
type KafkaServiceInterface interface {
	PublishAnalysisJob(message interface{}) error
	Close() error
}

type AnalysisService struct {
	db           *gorm.DB
	config       *config.Config
	kafkaService KafkaServiceInterface
}

func NewAnalysisService(db *gorm.DB, cfg *config.Config, kafkaService KafkaServiceInterface) *AnalysisService {
	return &AnalysisService{
		db:           db,
		config:       cfg,
		kafkaService: kafkaService,
	}
}

// AnalysisJobRequest represents the request to start analysis
type AnalysisJobRequest struct {
	TranscriptID uuid.UUID `json:"transcript_id" binding:"required"`
}

// AnalysisJobResponse represents the job creation response
type AnalysisJobResponse struct {
	JobID        uuid.UUID `json:"job_id"`
	TranscriptID uuid.UUID `json:"transcript_id"`
	Status       string    `json:"status"`
	Message      string    `json:"message"`
}

// JobStatusResponse represents the job status polling response
type JobStatusResponse struct {
	JobID        uuid.UUID  `json:"job_id"`
	TranscriptID uuid.UUID  `json:"transcript_id"`
	Status       string     `json:"status"` // pending, processing, completed, failed
	CreatedAt    time.Time  `json:"created_at"`
	CompletedAt  *time.Time `json:"completed_at,omitempty"`
	ErrorMessage *string    `json:"error_message,omitempty"`
}

// AnalysisResultsResponse represents complete analysis results
type AnalysisResultsResponse struct {
	ID                 uuid.UUID                `json:"id"`
	JobID              uuid.UUID                `json:"job_id"`
	TranscriptID       uuid.UUID                `json:"transcript_id"`
	Status             string                   `json:"status"`
	Summary            *string                  `json:"summary,omitempty"`
	Takeaways          []string                 `json:"takeaways,omitempty"`
	FactChecks         []FactCheckResultResponse `json:"fact_checks"`
	CreatedAt          time.Time                `json:"created_at"`
	CompletedAt        *time.Time               `json:"completed_at,omitempty"`
	TranscriptFilename *string                  `json:"transcript_filename,omitempty"`
	TranscriptTitle    *string                  `json:"transcript_title,omitempty"`
}

// FactCheckResultResponse represents individual fact-check results
type FactCheckResultResponse struct {
	ID         uuid.UUID `json:"id"`
	Claim      string    `json:"claim"`
	Verdict    string    `json:"verdict"`
	Confidence float64   `json:"confidence"`
	Evidence   *string   `json:"evidence,omitempty"`
	Sources    []string  `json:"sources,omitempty"`
	CheckedAt  time.Time `json:"checked_at"`
}

// KafkaMessage represents the message sent to Kafka
type KafkaMessage struct {
	JobID        uuid.UUID `json:"job_id"`
	TranscriptID uuid.UUID `json:"transcript_id"`
	CreatedAt    time.Time `json:"created_at"`
}

// CreateAnalysisJob creates a new analysis job
func (s *AnalysisService) CreateAnalysisJob(req *AnalysisJobRequest, correlationID string) (*AnalysisJobResponse, error) {
	log := logger.WithCorrelationID(correlationID)

	// Verify transcript exists
	var transcript models.Transcript
	if err := s.db.Where("id = ?", req.TranscriptID).First(&transcript).Error; err != nil {
		if err == gorm.ErrRecordNotFound {
			log.WithField("transcript_id", req.TranscriptID).Error("Transcript not found for analysis")
			return nil, fmt.Errorf("transcript %s not found", req.TranscriptID)
		}
		logger.LogErrorWithStackAndCorrelation(err, correlationID, map[string]interface{}{
			"transcript_id": req.TranscriptID,
			"operation":     "find_transcript_for_analysis",
		})
		return nil, fmt.Errorf("failed to find transcript: %w", err)
	}

	// Create analysis record
	analysis := &models.AnalysisResult{
		TranscriptID: req.TranscriptID,
		JobID:        uuid.New(),
		Status:       "pending",
	}

	if err := s.db.Create(analysis).Error; err != nil {
		logger.LogErrorWithStackAndCorrelation(err, correlationID, map[string]interface{}{
			"transcript_id": req.TranscriptID,
			"job_id":        analysis.JobID,
			"operation":     "create_analysis_job",
		})
		return nil, fmt.Errorf("failed to create analysis job: %w", err)
	}

	// Send message to Kafka
	message := KafkaMessage{
		JobID:        analysis.JobID,
		TranscriptID: analysis.TranscriptID,
		CreatedAt:    analysis.CreatedAt,
	}

	if err := s.kafkaService.PublishAnalysisJob(message); err != nil {
		logger.LogErrorWithStackAndCorrelation(err, correlationID, map[string]interface{}{
			"job_id":        analysis.JobID,
			"transcript_id": req.TranscriptID,
			"operation":     "publish_analysis_job_kafka",
		})
		// Update status to failed
		s.db.Model(analysis).Update("status", "failed")
		return nil, fmt.Errorf("failed to queue analysis job: %w", err)
	}

	log.WithFields(map[string]interface{}{
		"job_id":        analysis.JobID,
		"transcript_id": req.TranscriptID,
		"analysis_id":   analysis.ID,
	}).Info("Analysis job created")

	return &AnalysisJobResponse{
		JobID:        analysis.JobID,
		TranscriptID: req.TranscriptID,
		Status:       "pending",
		Message:      "Analysis job created and queued for processing",
	}, nil
}

// GetJobStatus returns the status of an analysis job
func (s *AnalysisService) GetJobStatus(jobID uuid.UUID, correlationID string) (*JobStatusResponse, error) {
	log := logger.WithCorrelationID(correlationID)

	var analysis models.AnalysisResult
	if err := s.db.Where("job_id = ?", jobID).First(&analysis).Error; err != nil {
		if err == gorm.ErrRecordNotFound {
			log.WithField("job_id", jobID).Error("Analysis job not found")
			return nil, fmt.Errorf("analysis job %s not found", jobID)
		}
		logger.LogErrorWithStackAndCorrelation(err, correlationID, map[string]interface{}{
			"job_id":    jobID,
			"operation": "get_job_status",
		})
		return nil, fmt.Errorf("failed to get job status: %w", err)
	}

	log.WithFields(map[string]interface{}{
		"job_id":      jobID,
		"status":      analysis.Status,
		"analysis_id": analysis.ID,
	}).Info("Retrieved job status")

	return &JobStatusResponse{
		JobID:        analysis.JobID,
		TranscriptID: analysis.TranscriptID,
		Status:       analysis.Status,
		CreatedAt:    analysis.CreatedAt,
		CompletedAt:  analysis.CompletedAt,
		ErrorMessage: analysis.ErrorMessage,
	}, nil
}

// GetAnalysisResults returns complete analysis results
func (s *AnalysisService) GetAnalysisResults(analysisID uuid.UUID, correlationID string) (*AnalysisResultsResponse, error) {
	log := logger.WithCorrelationID(correlationID)

	// Join with transcript to get filename and metadata
	var analysis models.AnalysisResult
	var transcript models.Transcript
	
	if err := s.db.Where("id = ?", analysisID).First(&analysis).Error; err != nil {
		if err == gorm.ErrRecordNotFound {
			log.WithField("analysis_id", analysisID).Error("Analysis not found")
			return nil, fmt.Errorf("analysis %s not found", analysisID)
		}
		logger.LogErrorWithStackAndCorrelation(err, correlationID, map[string]interface{}{
			"analysis_id": analysisID,
			"operation":   "get_analysis",
		})
		return nil, fmt.Errorf("failed to get analysis: %w", err)
	}

	if err := s.db.Where("id = ?", analysis.TranscriptID).First(&transcript).Error; err != nil {
		logger.LogErrorWithStackAndCorrelation(err, correlationID, map[string]interface{}{
			"transcript_id": analysis.TranscriptID,
			"analysis_id":   analysisID,
			"operation":     "get_transcript_for_analysis",
		})
		return nil, fmt.Errorf("failed to get transcript: %w", err)
	}

	// Load fact checks
	var factChecks []models.FactCheck
	if err := s.db.Where("analysis_id = ?", analysisID).Find(&factChecks).Error; err != nil {
		logger.LogErrorWithStackAndCorrelation(err, correlationID, map[string]interface{}{
			"analysis_id": analysisID,
			"operation":   "load_fact_checks",
		})
		return nil, fmt.Errorf("failed to load fact checks: %w", err)
	}

	// Convert fact checks to response format
	factCheckResponses := make([]FactCheckResultResponse, len(factChecks))
	for i, fc := range factChecks {
		var sources []string
		if fc.Sources != nil {
			json.Unmarshal(fc.Sources, &sources)
		}
		
		factCheckResponses[i] = FactCheckResultResponse{
			ID:         fc.ID,
			Claim:      fc.Claim,
			Verdict:    fc.Verdict,
			Confidence: fc.Confidence,
			Evidence:   fc.Evidence,
			Sources:    sources,
			CheckedAt:  fc.CheckedAt,
		}
	}

	// Convert takeaways from JSON
	var takeaways []string
	if analysis.Takeaways != nil {
		json.Unmarshal(analysis.Takeaways, &takeaways)
	}

	// Extract title from transcript metadata if available
	var transcriptTitle *string
	if transcript.TranscriptMetadata != nil {
		var metadata map[string]interface{}
		if err := json.Unmarshal(transcript.TranscriptMetadata, &metadata); err == nil {
			if title, ok := metadata["title"].(string); ok {
				transcriptTitle = &title
			}
		}
	}

	log.WithFields(map[string]interface{}{
		"analysis_id":        analysisID,
		"status":             analysis.Status,
		"fact_checks_count":  len(factChecks),
	}).Info("Retrieved analysis results")

	return &AnalysisResultsResponse{
		ID:                 analysis.ID,
		JobID:              analysis.JobID,
		TranscriptID:       analysis.TranscriptID,
		Status:             analysis.Status,
		Summary:            analysis.Summary,
		Takeaways:          takeaways,
		FactChecks:         factCheckResponses,
		CreatedAt:          analysis.CreatedAt,
		CompletedAt:        analysis.CompletedAt,
		TranscriptFilename: &transcript.Filename,
		TranscriptTitle:    transcriptTitle,
	}, nil
}

// ListAnalysisResults returns paginated list of analysis results
func (s *AnalysisService) ListAnalysisResults(page, perPage int) ([]*AnalysisResultsResponse, int64, error) {
	var results []struct {
		models.AnalysisResult
		TranscriptFilename string `json:"transcript_filename"`
	}
	var total int64

	offset := (page - 1) * perPage

	// Count total
	if err := s.db.Model(&models.AnalysisResult{}).Count(&total).Error; err != nil {
		logger.LogErrorWithStack(err, map[string]interface{}{
			"operation": "count_analysis_results",
			"page":      page,
			"per_page":  perPage,
		})
		return nil, 0, fmt.Errorf("failed to count analysis results: %w", err)
	}

	// Get results with transcript filename
	if err := s.db.
		Table("analysis_results").
		Select("analysis_results.*, transcripts.filename as transcript_filename").
		Joins("JOIN transcripts ON analysis_results.transcript_id = transcripts.id").
		Order("analysis_results.created_at DESC").
		Offset(offset).
		Limit(perPage).
		Scan(&results).Error; err != nil {
		logger.LogErrorWithStack(err, map[string]interface{}{
			"operation": "get_analysis_results_list",
			"page":      page,
			"per_page":  perPage,
			"offset":    offset,
		})
		return nil, 0, fmt.Errorf("failed to get analysis results: %w", err)
	}

	// Convert to response format
	responses := make([]*AnalysisResultsResponse, len(results))
	for i, result := range results {
		// Load fact checks for this analysis
		var factChecks []models.FactCheck
		s.db.Where("analysis_id = ?", result.ID).Find(&factChecks)

		factCheckResponses := make([]FactCheckResultResponse, len(factChecks))
		for j, fc := range factChecks {
			var sources []string
			if fc.Sources != nil {
				json.Unmarshal(fc.Sources, &sources)
			}
			
			factCheckResponses[j] = FactCheckResultResponse{
				ID:         fc.ID,
				Claim:      fc.Claim,
				Verdict:    fc.Verdict,
				Confidence: fc.Confidence,
				Evidence:   fc.Evidence,
				Sources:    sources,
				CheckedAt:  fc.CheckedAt,
			}
		}

		// Convert takeaways from JSON
		var takeaways []string
		if result.Takeaways != nil {
			json.Unmarshal(result.Takeaways, &takeaways)
		}

		responses[i] = &AnalysisResultsResponse{
			ID:                 result.ID,
			JobID:              result.JobID,
			TranscriptID:       result.TranscriptID,
			Status:             result.Status,
			Summary:            result.Summary,
			Takeaways:          takeaways,
			FactChecks:         factCheckResponses,
			CreatedAt:          result.CreatedAt,
			CompletedAt:        result.CompletedAt,
			TranscriptFilename: &result.TranscriptFilename,
		}
	}

	return responses, total, nil
}

// UpdateJobStatus updates the status of an analysis job (matches Python def update_job_status)
func (s *AnalysisService) UpdateJobStatus(jobID uuid.UUID, status string, errorMessage string) error {
	var analysis models.AnalysisResult
	if err := s.db.Where("job_id = ?", jobID).First(&analysis).Error; err != nil {
		logger.LogErrorWithStack(err, map[string]interface{}{
			"job_id":    jobID,
			"operation": "find_job_for_status_update",
		})
		return err
	}

	analysis.Status = status
	if errorMessage != "" {
		analysis.ErrorMessage = &errorMessage
	}
	if status == "completed" || status == "failed" {
		now := time.Now()
		analysis.CompletedAt = &now
	}

	if err := s.db.Save(&analysis).Error; err != nil {
		logger.LogErrorWithStack(err, map[string]interface{}{
			"job_id":      jobID,
			"analysis_id": analysis.ID,
			"status":      status,
			"operation":   "save_job_status_update",
		})
		return err
	}

	logger.Log.WithFields(map[string]interface{}{
		"job_id": jobID,
		"status": status,
		"analysis_id": analysis.ID,
	}).Info("Updated job status")

	return nil
}